text/0000-orderer-snapshot.md

```diff
@@ -0,0 +1,263 @@
+3. Rename "Channel participation API Configuration" section in the OSN YAML config file to "Channel management API Configuration", rename accordingly all the config settings
+4. Add two new snapshot operations to the REST API definitions of **/v1/mgmt/channels** REST API
+5. Add support of two new snapshot operations to the CLI tool *osnadmin*, specifically to the command *osnadmin channel* and the new command *osnadmin snapshot*
+6. Deprecate *osnadmin channel join* command. This will be the only command using deprecated **/v1/participation/channels**, all the rest of the commands including inherited *osnadmin channel remove* and *osnadmin channel list* must use internally new **/v1/mgmt/channels** endpoint.
+
+## Snapshot content
+
+We propose the OSN snapshot to follow the same logical structure as peer snapshot, that is to comprise a directory with a number of files in it, one of the files being a metadata JSON file with the checksums of other files and other metadata. But here the similarity ends, as the content of the files will be drastically different.
+Here is the content of the OSN snapshot:
+1. Snapshot version - to allow for the future extensions. Current version is 1.
+2. Channel ID.
+3. Last config block for the channel.
+4. lastBlockNum - the number of the last pruned block. This can be both behind the last config block and ahead of it.
+5. lastBlockHash - the hash of the last pruned block.
+6. prevBlockHash - technically not required, but this will make internal structure BlockchainInfo complete.
+
+Version 1 of the OSN snapshot described in this RFC defines two files as the content of the snapshot directory:
+1. Last config block for the channel. Name: "lastConfig.block"
+2. Metadata file containing all the rest of the snapshot content. Name: "metadata.json"
+
+Unlike peer snapshot the OSN snapshot does not contain state DB, private data configs, the list of TxIDs and commit hash - as the OSN is purposefully designed to be unaware of these. So these two kinds of snapshots (peer snapshot and OSN snapshot) bear resemblance only on the conceptual level, being completely different in the content.
+
+## Implementation considerations
+
+When a channel is joined from a snapshot by "channel joinbysnapshot" command, we have to save the following info on the block storage:
+1. Last config block from snapshot
+2. BootstrappingSnapshotInfo
+
+For BootstrappingSnapshotInfo persistence the implementation is already there, but for the last config block a new persistent unit of storage (a file in case of file-based ledger) should be created. Additionally all the procedures looking up the last config block should be modified to return the last config block from the snapshot in case the last config block is not found in the block ledger itself.
```

Comment:
> Additionally all the procedures looking up the last config block should be modified to return the last config block from the snapshot in case the last config block is not found in the block ledger itself.

I don't quite follow this.  Typically, to find the last config block from an orderer a client pulls the latest block, looks at its metadata to find a pointer to the latest config block, and then pulls that block.  Are you suggesting we augment the Deliver service with the ability to pull the original config block even if it is not in the ledger?

Reply:
The text in RFC is about searching for the last config block locally, in the local ledger of an OSN. It is not about the remote block fetching.
Here are the couple of places (not exhaustive list!) where the search in local ledger should be modified to also account for the possibility that the last config block is not in the ledger, but in a bootstrapping storage unit (in a separate file):
https://github.com/hyperledger/fabric/blob/4cfc722d0ffef35a4ac8b0ebe17eea4dcc3b1018/orderer/common/follower/follower_chain.go#L552
https://github.com/hyperledger/fabric/blob/4cfc722d0ffef35a4ac8b0ebe17eea4dcc3b1018/orderer/common/cluster/util.go#L648

Also I have clarified the text in the RFC:
a916490fd6fb66e5c5903de8a062ee4cc3bb488a